bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/operations/OperationHistoryActionHandler.java

```diff
@@ -298,7 +303,32 @@ public final void run() {
 			if (getOperation() instanceof IAdvancedUndoableOperation2) {
 				runInBackground = ((IAdvancedUndoableOperation2) getOperation()).runInBackground();
 			}
-			progressDialog.run(runInBackground, true, runnable);
+			if (runInBackground) {
+				progressDialog.run(runInBackground, true, runnable);
+			} else {
+				final IJobManager manager = Job.getJobManager();
+				final ISchedulingRule rule = (getOperation() instanceof ISchedulableOperation)
```

Comment:
I wonder if that can work as proposed. 

An undo operation may want to perform different tasks and acquire multiple rules for them, so first acquire rule on one resource and later on another etc.

Another reason I'm asking is: an **existing** "generic" operation that can be extended can't use this new 
`ISchedulableOperation` interface to force the code here use the rule, because **if** it can be extended, the 3rd party code may don't want it to happen, so it will break existing code again.

May be it would be good to see what is the actual change planned on the client side (in JDT? LTK?).

Reply:
see https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/19
it's not a secret.

Reply:
> see [eclipse-jdt/eclipse.jdt.ui#19](https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/19) it's not a secret.

Thanks, I will look into. Would be easier for all if that would be mentioned in this PR somewhere.

Reply:
It was autolinked and for sure you got a notification about that. Probably you just overlooked it.